packages/delivery-expo/delivery.js

```diff
@@ -0,0 +1,138 @@
+const payload = require('@bugsnag/core/lib/json-payload')
+const { isoDate } = require('@bugsnag/core/lib/es-utils')
+const UndeliveredPayloadQueue = require('./queue')
+const NetworkStatus = require('./network-status')
+const RedeliveryLoop = require('./redelivery')
+
+module.exports = (client, fetch = global.fetch) => {
+  const networkStatus = new NetworkStatus()
+
+  const send = (url, opts, cb) => {
+    fetch(url, opts)
+      .then(response => {
+        if (response.ok) return response.text()
+        const err = new Error(`Bad status code from API: ${response.status}`)
+        err.isRetryable = isRetryable(response.status)
+        return Promise.reject(err)
+      })
+      .then(() => cb(null))
+      .catch(err => cb(err))
+  }
+
+  const logError = e => client._logger.error('Error redelivering payload', e)
+
+  const enqueue = async (payloadKind, failedPayload) => {
+    client._logger.info(`Writing ${payloadKind} payloads to cache`)
+    await queues[payloadKind].enqueue(failedPayload, logError)
+    if (networkStatus.isConnected) queueConsumers[payloadKind].start()
+  }
+
+  const onerror = async (err, failedPayload, payloadKind, cb) => {
+    client._logger.error(`${payloadKind} failed to send…\n${(err && err.stack) ? err.stack : err}`, err)
+    if (failedPayload && err.isRetryable !== false) enqueue(payloadKind, failedPayload)
+    cb(err)
+  }
+
+  const { queues, queueConsumers } = initRedelivery(networkStatus, logError, send)
+
+  return {
+    sendReport: (report, cb = () => {}) => {
+      const url = client.config.endpoints.notify
+
+      let body, opts
+      try {
+        body = payload.report(report, client.config.filters)
+        opts = {
+          method: 'POST',
+          headers: {
+            'Content-Type': 'application/json',
+            'Bugsnag-Api-Key': report.apiKey || client.config.apiKey,
+            'Bugsnag-Payload-Version': '4',
+            'Bugsnag-Sent-At': isoDate()
+          },
+          body
+        }
+        if (!networkStatus.isConnected || (report._handledState && report._handledState.attemptDelivery === false)) {
+          enqueue('report', { url, opts })
+          return cb(null)
+        }
+        client._logger.info(`Sending report ${report.errorClass}: ${report.errorMessage}`)
+        send(url, opts, err => {
+          if (err) return onerror(err, { url, opts }, 'report', cb)
+          cb(null)
+        })
+      } catch (e) {
+        onerror(e, { url, opts }, 'report', cb)
+      }
+    },
+
+    sendSession: (session, cb = () => {}) => {
+      const url = client.config.endpoints.sessions
+
+      let body, opts
+      try {
+        body = payload.session(session, client.config.filters)
+        opts = {
+          method: 'POST',
+          headers: {
+            'Content-Type': 'application/json',
+            'Bugsnag-Api-Key': client.config.apiKey,
```

Comment:
I did consider this and I think it's fine. All errors before the change would go the old project and all errors after the change would go to the new project. That seems reasonable to me.

Plus if you made a change – say moving from on prem to bugsnag.com, removing some PII from from the payload in the process – it would be surprising if some old errors with all the PII went to bugsnag.com.

Reply:
The scenario I'm thinking of would be:

1. error captured and stored on disk
2. API key changes before next launch
3. stored error is sent using the new API key, rather than the old one

Reply:
That wouldn't be the case – this block of code is only ever entered once for an error report. When it is taken from disk it goes via the lower level `send()` function in this file, not this `sendReport()` function.